### ISSUE TITLE
React UI: insufficient significant digits in Y axis labels over small range

### ISSUE THREAD

**Issue by candlerb** [prometheus/prometheus#9204]
**What did you do?**

In the PromQL React UI, gave a PromQL expression where the Y values are over a small range, then view as a graph.

To replicate: e.g. `foo < 0.10 > 0.096` or `foo < 0.11 > 0.096`

**What did you expect to see?**

Each Y-axis label should be distinct, with sufficient significant digits to distinguish them.  e.g.

* 0.096
* 0.097
* 0.098
* 0.099
* 0.100
* 0.101

**What did you see instead? Under which circumstances?**

In the following graph, the data points range from 0.0965 to 0.0999:

![image](https://user-images.githubusercontent.com/44789/129471355-2fc6c307-5737-4e55-bfc2-0d3fbad67d97.png)

I believe the Y axis label positions actually correspond to 0.0960, 0.0965, 0.0970 ... 0.0100, 0.0105, but all are showing as 0.10.

With values from 0.0965 to 0.11:

![image](https://user-images.githubusercontent.com/44789/129471416-1ba72c16-cbe4-48ac-8b78-e181caf73e37.png)

In this case I believe the Y axis positions correspond to 0.096, 0.098, 0.100, 0.102, ... 0.110, 0.112; but half of them show as 0.10 and half as 0.11

(Real-world use case which triggered this: alerts set at disks having less than 10% free space, then viewing the graph of the alerting expression when the values are only just dipping below and above 10%)

**Workaround**

Manually add `* 100` (or other scale as required) to the expression you are graphing.

![image](https://user-images.githubusercontent.com/44789/129471876-ea428237-cb44-412e-9fa8-e9ce7f97919e.png)

**Reference**

Possibly relates to old issue #3084

**Environment**

* System information:

    ```
    Linux 4.15.0-153-generic x86_64
    ```

* Prometheus version:

    ```
    prometheus, version 2.28.1 (branch: HEAD, revision: b0944590a1c9a6b35dc5a696869f75f422b107a1)
      build user:       root@2915dd495090
      build date:       20210701-15:20:10
      go version:       go1.16.5
      platform:         linux/amd64
    ```

* Alertmanager version:

    n/a

* Prometheus configuration file:

    n/a

* Alertmanager configuration file:

    n/a

* Logs:

    n/a

**Comment by roidelapluie** [comment]
cc @juliusv

**Comment by krajorama** [comment]
Hello from the bug scrub: is this relevant for the new library for the UI @juliusv ?

**Comment by juliusv** [comment]
Graphing is the major piece still missing in the new UI, so we'll see :)

**Comment by juliusv** [comment]
Yeah, the new UI is using the same tick formatter as before: https://github.com/prometheus/prometheus/blob/52fe4cc4eeeb98ab15086501412f9daf05835d71/web/ui/mantine-ui/src/pages/query/uPlotChartHelpers.ts#L8-L54

If someone wants to change it, just make sure that the Y-axis label space allocator in https://github.com/prometheus/prometheus/blob/52fe4cc4eeeb98ab15086501412f9daf05835d71/web/ui/mantine-ui/src/pages/query/uPlotChartHelpers.ts#L194-L224 works well enough to allocate enough space for longer tick labels. That part is unfortunately quite hacky with uPlot (as are many things with uPlot, unfortunately), copied from a uPlot demo.

**Comment by flying-musk** [comment]
Hi, I’m interested in working on this issue.

I noticed this issue is old, so before I start, I just wanted to confirm whether this is still something the maintainers would be open to. If so, I’d be happy to work on it. Could this issue be assigned to me?

**Comment by juliusv** [comment]
@flying-musk Yes, should still be fine, gonna assign you then :)

**Comment by flying-musk** [comment]
@juliusv, I opened a PR for this issue here and updated the Y-axis tick formatter to increase decimal precision when the default formatting would produce duplicate labels.

I also kept the existing SI-prefix formatting behavior and manually tested the small-range cases mentioned in the PR description.

Whenever you have time, I’d appreciate any feedback. Thank you!
